bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/internal/WebClientFactoryImpl.java

```diff
@@ -96,30 +102,55 @@ protected void deactivate() {
             try {
                 commonHttpClient.stop();
             } catch (Exception e) {
-                logger.error("error while stopping shared jetty http client", e);
+                logger.error("error while stopping shared Jetty http client", e);
                 // nothing else we can do here
             }
             commonHttpClient = null;
-            threadPool = null;
-            logger.debug("jetty shared http client stopped");
+            logger.debug("Jetty shared http client stopped");
         }
+        if (commonWebSocketClient != null) {
+            try {
+                commonWebSocketClient.stop();
+            } catch (Exception e) {
+                logger.error("error while stopping shared Jetty web socket client", e);
+                // nothing else we can do here
+            }
+            commonWebSocketClient = null;
+            logger.debug("Jetty shared web socket client stopped");
+        }
+        threadPool = null;
     }
 
     @Override
-    public HttpClient createHttpClient(@NonNull String consumerName, @NonNull String endpoint) {
+    public HttpClient createHttpClient(String consumerName, String endpoint) {
         Objects.requireNonNull(endpoint, "endpoint must not be null");
-        logger.debug("httpClient for endpoint {} requested", endpoint);
+        logger.debug("http client for endpoint {} requested", endpoint);
         checkConsumerName(consumerName);
         return createHttpClientInternal(consumerName, endpoint, false);
     }
 
+    @Override
+    public WebSocketClient createWebSocketClient(String consumerName, String endpoint) {
+        Objects.requireNonNull(endpoint, "endpoint must not be null");
+        logger.debug("web socket client for endpoint {} requested", endpoint);
+        checkConsumerName(consumerName);
+        return createWebSocketClientInternal(consumerName, endpoint, false);
+    }
+
     @Override
     public HttpClient getCommonHttpClient() {
         initialize();
-        logger.debug("shared httpClient requested");
+        logger.debug("shared http client requested");
         return commonHttpClient;
     }
 
+    @Override
+    public WebSocketClient getCommonWebSocketClient() {
+        initialize();
+        logger.debug("shared web socket client requested");
+        return commonWebSocketClient;
+    }
+
     private void getConfigParameters(Map<String, Object> parameters) {
```

Comment:
@maggu2810 does OSGI give us the guarantee this is never null? In that case I can include killing the the null-check on line 173 in my pending PR.

<img width="805" alt="screenshot 2018-09-28 at 15 26 34" src="https://user-images.githubusercontent.com/2625279/46211101-f0c4a380-c332-11e8-9a32-876d01a87230.png">

I can also imagine that keys will not be guaranteed to be there so something should be annotated such that its aware that the map does not contain every possible key.

Is that just a matter of:

```java
    @Activate
    protected void activate(Map<String, @Nullable Object> parameters) {
        logger.trace("Activated");
        getConfigParameters(parameters);
    }

    @Modified
    protected void modified(Map<String, @Nullable Object> parameters) {
        getConfigParameters(parameters);
        if (threadPool != null) {
            threadPool.setMinThreads(minThreadsShared);
            threadPool.setMaxThreads(maxThreadsShared);
            threadPool.setIdleTimeout(keepAliveTimeoutShared * 1000);
        }
    }
```

Thanks!

Reply:
AFAIK the map itself will not be null.
AFAIK the key and the values will not be null.

But it is not explicitly stated in the specification.
It is IMHO implicit stated by the combination of some sentences in the specification.

IIRC we already commented this in another issue some weeks ago.
@kaikreuzer would like to use explicit null checks as long as there is nothing explicit stated.

@kaikreuzer Do you remember of the issue we comment about that?

Reply:
Does that mean that technically the default value should never be used?

Reply:
You refer to the map given to the activate and modified method?
Where do you define such a default value?

Reply:
I refer to the warnings in the `getConfigParameter` (see screenshot above), there the compiler claims that the `Map` will always return a `NonNull` value for any key we throw at it. But we do a check to confirm that if the value returned from the `Map` is `null` then we return the `defaultValue` instead.

Reply:
@martinvw Perhaps I should write a Wiki entry some time as I comment such topics such often already (sorry, no offend to you).

I it caused by missing EEAs for the Java classes.

The method `V get(K key)` is not annotated (we could provide EEAs) and so `V` is interpret as `@NonNull Object` (depend on the annotation of `V`). That's wrong as the method signature should be `@Nullable V get(K key)`. The method can potential always return null if e.g. `key` is not part of the map.
So, we should use EEAs or have to live with that warnings (as long as they are not disabled).

Reply:
Sorry, I remembered that the code guideline states:

> Code should not show any warnings. Warnings that cannot be circumvented should be suppressed by using the @SuppressWarnings annotation.

So I should suppress the warning over the null check. The additional Nullable felt way more clean to me, but let me suppress it then

Reply:
> @kaikreuzer Do you remember of the issue we comment about that?

It is https://github.com/eclipse/smarthome/pull/6095.

> Perhaps I should write a Wiki entry some time as I comment such topics such often already (sorry, no offend to you).

You actually already almost did in the still open https://github.com/eclipse/smarthome/issues/5908, which discusses exactly that issue. This happened during my holidays, so I didn't involve myself in it and just saw it now when searching for our past discussion.

> The additional Nullable felt way more clean to me, but let me suppress it then

Yeah, I felt the same, but I am now also adding `SuppressWarnings` as an ugly workaround. I still think the real issue is that it was decided to use `@NonNullByDefault` on class level, which is simply a false statement.